pkg/bigquery/sanitize.go

```diff
@@ -0,0 +1,99 @@
+)
+
+const tableNameMaxUTF8Bytes = 1024
+
+// Start with lowercase, middle is lowercase, number or hyphen, cannot end in
+// hyphen. 6-30 characters in length (start, 4-28 middle, end).
+var projectIDMatcher = regexp.MustCompile(`^[a-z][a-z0-9\-]{4,28}[a-z0-9]$`)
+
+// Lowercase and uppercase letters and underscores. Max 1024 characters.
+// regexp only allows 1000 repetitions, so had to manually repeat.
+const datasetIDRegex = `^[a-zA-Z0-9_]{1,512}[a-zA-Z0-9_]{0,512}$`
+
+var datasetIDMatcher = regexp.MustCompile(datasetIDRegex)
+
+// Unicode characters in category L (letter), M (mark), N (number),
+// Pc (connector, including underscore), Pd (dash), Zs (space). Max 1024
+// characters. 1024 is actually UTF-8 bytes from experimentation.
+// UTF-8 length check will be done separately.
+// regexp only allows 1000 repetitions, so had to manually repeat.
+const tableNameRegex = `^[\p{L}\p{M}\p{N}\p{Pc}\p{Pd}\p{Zs}]{1,512}[\p{L}\p{M}\p{N}\p{Pc}\p{Pd}\p{Zs}]{0,512}$`
+
+var tableNameMatcher = regexp.MustCompile(tableNameRegex)
+
+// Returns nil if provided string is a valid GCP project id, else returns an
+// error. Based on rules defined in
+// [https://cloud.google.com/resource-manager/docs/creating-managing-projects].
+// Identifiers cannot be replaced via parameters in BigQuery, so interpolation
+// is a necessry evil.
+// In most cases, setting the default GCP project in the client's Query object
+// is preferable as it avoids the injection risk entirely.
+// Does not check for restricted strings such as google, null, etc.
+func ValidateGCPProjectID(projectID string) error {
+	if !projectIDMatcher.MatchString(projectID) {
+		return fmt.Errorf("invalid GCP project id")
+	}
+	return nil
+}
+
+// Returns nil if provided string is a valid dataset id, else returns an
+// error. Based on rules defined in
+// [https://cloud.google.com/bigquery/docs/datasets#dataset-naming].
+// Identifiers cannot be replaced via parameters in BigQuery, so interpolation
+// is a necessry evil.
+// In most cases, setting the default dataset in the client's Query object
+// is preferable as it avoids the injection risk entirely.
+// Does not check for restricted strings such as google, null, etc.
+func ValidateDatasetID(datasetID string) error {
+	if !datasetIDMatcher.MatchString(datasetID) {
+		return fmt.Errorf("invalid dataset id")
+	}
+	return nil
+}
+
+// Returns nil if provided string is a valid table name, else returns an
+// error. Based on rules defined in
+// [https://cloud.google.com/bigquery/docs/tables#table_naming].
+// Identifiers cannot be replaced via parameters in BigQuery, so interpolation
+// is a necessry evil.
+// Hardcoding table names may be preferred to avoid any injection risk.
+// Unclear if these rules are valid for all external tables.
+// Does not check for restricted strings such as google, null, etc.
+func ValidateTableName(tableName string) error {
+	// TODO: is this really necessary, seems like regexp may do this for free.
```

Comment:
TODOs should be imperative (e.g. TODO: `Do this when blah blah`). See internal resources for TODO style. Otherwise, just remove this altogether if there's not really any need for it.

Reply:
Thanks, removed todo but left the check as its cheap